README.md
[![Badge for version for Visual Studio Code extension MPearon.vsce-show-triggerwords](https://vsmarketplacebadge.apphb.com/version/MPearon.vsce-show-triggerwords.svg?color=darkred&style=flat&logo=visual-studio-code)](https://marketplace.visualstudio.com/items?itemName=MPearon.vsce-show-triggerwords&wt.mc_id=MPearon.vsce-show-triggerwords) [![Rating](https://vsmarketplacebadge.apphb.com/rating/MPearon.vsce-show-triggerwords.svg?color=darkred&style=flat)](https://marketplace.visualstudio.com/items?itemName=MPearon.vsce-show-triggerwords&wt.mc_id=MPearon.vsce-show-triggerwords) [![Installs](https://vsmarketplacebadge.apphb.com/installs-short/MPearon.vsce-show-triggerwords.svg?color=darkred&style=flat)](https://marketplace.visualstudio.com/items?itemName=MPearon.vsce-show-triggerwords&wt.mc_id=MPearon.vsce-show-triggerwords) [![Downloads](https://vsmarketplacebadge.apphb.com/downloads-short/MPearon.vsce-show-triggerwords.svg?color=darkred&style=flat)](https://marketplace.visualstudio.com/items?itemName=MPearon.vsce-show-triggerwords&wt.mc_id=MPearon.vsce-show-triggerwords) [![Live Share](https://img.shields.io/badge/Live_Share-enabled-8F80CF.svg?color=darkred&style=flat&logo=visual-studio-code)](https://visualstudio.microsoft.com/services/live-share/?wt.mc_id=MPearon.vsce-show-triggerwords) [![The MIT License](https://img.shields.io/badge/license-MIT-orange.svg?color=darkred&style=flat)](http://opensource.org/licenses/MIT)

# Purpose
This extension's purpose is to help emphasize user-specified "trigger words" in certain file types.

# Features

## Highlighting
### Whole-line Highlighting
- If a trigger word is detected, the entire line is highlighted lightly, and the matched word is emphasized heavily.
### Automatic and Manual Triggering
- Files with '.log' extension are automatically highlighted when the file is opened and "live" as the file is modified
- Other files can be manually highlighted by triggering the 'Show Trigger Words' command from the Command Pallette
### Highlighting Details
- "Error", "Err", "Failure", "Fail", "Critical" and "Crit" are highlighted red.
- "Warning" and "Warn" are highlighted yellow.
- "Information" and "Info" are highlighted blue.
- "Success", "Successful" and "Succeeded" are highlighted green.

## Parsing
Not available for initial release.

## Future Features
### Highlighting
- [x] User-specified words/RegEx
- [x] User-specified colors
- [ ] Whole-line highlighting
	- [x] Entire line highlighted light ([25a691f](https://github.com/mpearon/PUB-vsce.show-TriggerWords/commit/25a691fb9d97f55b2917196d0d742de1c41d46ef))
	- [x] Matched word highlighted dark ([25a691f](https://github.com/mpearon/PUB-vsce.show-TriggerWords/commit/25a691fb9d97f55b2917196d0d742de1c41d46ef))
	- [ ] Enable user toggle
- [ ] Optional Gutter markers
	- [ ] Color matched
### Parsing
- [ ] Provide canned RegEx to detect date, level and message.
- [ ] Optional user-supplied RegEx

# Extension Settings
No configurable settings at this time.

# Known Issues
To report issues, use [this link](https://github.com/mpearon/PUB-vsce.show-TriggerWords/issues).
- Comments are not ignored at this time.
- Switching between supported and non-supported languages can prevent highlighting updates.

# Release Notes
Reference [CHANGELOG](https://github.com/mpearon/PUB-vsce.show-TriggerWords/blob/master/CHANGELOG.md) for documentation about changes made to this repository

CHANGELOG.md
# v0.2.17 - 2021-03-31
- version bump from ```0.2.16``` to ```0.2.17```
- Merged dependabot vulnerability
- Updated CHANGELOG
# v0.2.16 - 2021-06-18
- version bump from ```0.2.15``` to ```0.2.16```
- Merged dependabot vulnerability
- Updated CHANGELOG
# v0.2.15 - 2021-04-12
- version bump from ```0.2.14``` to ```0.2.15```
- Merged dependabot vulnerability
# v0.2.14 - 2020-08-20
- Version bump from ```0.2.13``` to ```0.2.14```
# v0.2.13 - 2020-04-07
- Addressed minimist vulnerability
  ```
  npm install minimist
  npm update
  npm audit fix
  ```
- Set ```vscode.engine``` to ```^1.43.0``` 

# v0.2.12 - 2020-02-12
- Added provisions for contributions model

# v0.2.11 - 2020-01-23
- Updated package.json and package-lock.json to be npm-compliant

# v0.2.10 - 2020-01-22
- [Issue #8](#8): Removed requirement for there to characters proceeding or following the Trigger Word

# v0.2.9 - 2020-01-21
- Added provisions for live highlighting as the file is modified ([526004c](https://github.com/mpearon/PUB-vsce.show-TriggerWords/commit/526004c0e225cf314656eb1264b8ea150455a642))
- Prepared for user contributions
- Updated README
  - Features
    - Added verbiage to 'Automatic and Manual Triggering' header about live highlighting
    - Corrected color for 'Warning' under 'Highlighting Details' header
  - Known Issues
    - Removed entry for auto-updating

# v0.2.8 - 2020-01-21
- Added provisions to allow any file type to be highlighted via triggering 'Show Trigger Words' from Command Palette
- Updated README
  - Features
    - Restructured to use sub-headers instead of bullets
  - Known Issues
    - Removed '.log' language restriction entry
    - Added entry for auto-updating
- Corrected spelling issue in CHANGELOG entry for v0.1.0

# v0.2.7 - 2020-01-20
- Corrected README formatting issue

# v0.2.6 - 2020-01-20
- Adjusted badge location in README

# v0.2.5 - 2020-01-20
- Added badges to README

# v0.2.4 - 2020-01-20
- Adjusted branding
  - Icons
    - resources/altIcon-48.png
    - resources/altIcon-256.png
  - Gallery Info
    - icon

# v0.2.3 - 2020-01-20
- Added branding
  - Icons
    - resources/icon-48.png
    - resources/icon-256.png
  - Gallery Info
    - galleryBanner
    - icon
    - extensionKind
    - preview
- Corrected spelling in CHANGELOG

# v0.2.2 - 2020-01-20
- Corrected spelling issues in README

# v0.2.1 - 2020-01-20
- Corrected formatting issue in README

# v0.2.0 - 2020-01-20
- Added Whole-line Highlighting
- Added '.log' languageId restriction (temporary)
- Updated README
  - Updated to reflect complete goals
  - Added known issues

# v0.1.6 - 2020-01-20
- Added new goals to README

# v0.1.5 - 2020-01-20
- Corrected README and CHANGELOG formatting

# v0.1.4 - 2020-01-20
- Adding MIT licensing information

# v0.1.3 - 2020-01-19
- Reverted publisher name from 'Matthew Pearon' to 'MPearon'

# v0.1.2 - 2020-01-19
- Updated publisher name from 'MPearon' to 'Matthew Pearon'

# v0.1.1 - 2020-01-19
- Adjusted decorationOption RegEx entries to remove requirement that trigger word be first word on line

# v0.1.0 - 2020-01-19
- Added logic to loop through array of trigger words and their associated decoration options

# v0.0.0 - 2020-01-16
- Initial commit

src/extension.ts
import {
	window,
	commands,
	ExtensionContext,
	Range,
	workspace
} from 'vscode';

export function getColorSettings(){
	// Define the color options - will be moved to a separate file and be accessible via contributions 
	const color1Word = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.matchColor, color: 'black', fontWeight: 'bolder' };
	const color1Line = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.matchColor+'30', isWholeLine: true };
	const color2Word = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.matchColor, color: 'black', fontWeight: 'bolder' };
	const color2Line = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.matchColor+'30', isWholeLine: true };
	const color3Word = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.matchColor, color: 'black', fontWeight: 'bolder' };
	const color3Line = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.matchColor+'30', isWholeLine: true };
	const color4Word = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.matchColor, color: 'black', fontWeight: 'bolder' };
	const color4Line = { backgroundColor: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.matchColor+'30', isWholeLine: true };

	// Build the decorationTypes
	const color1LineType = window.createTextEditorDecorationType(color1Line);
	const color1WordType = window.createTextEditorDecorationType(color1Word);
	
	const color2LineType = window.createTextEditorDecorationType(color2Line);
	const color2WordType = window.createTextEditorDecorationType(color2Word);

	const color3LineType = window.createTextEditorDecorationType(color3Line);
	const color3WordType = window.createTextEditorDecorationType(color3Word);
	
	const color4LineType = window.createTextEditorDecorationType(color4Line);
	const color4WordType = window.createTextEditorDecorationType(color4Word);

	return {
		color1LineType,
		color2LineType,
		color3LineType,
		color4LineType,
		color1WordType,
		color2WordType,
		color3WordType,
		color4WordType
	}
};

export function activate(context: ExtensionContext) {
	let colors = getColorSettings()
	console.log('mpearon.vsce.show-triggerwords has been initialized');
	// No editor is open, cancel load
	if (!window.activeTextEditor)
		return;

	function clearDecorations(){
		colors.color1LineType.dispose()
		colors.color1WordType.dispose()
		colors.color2LineType.dispose()
		colors.color2WordType.dispose()
		colors.color3LineType.dispose()
		colors.color3WordType.dispose()
		colors.color4LineType.dispose()
		colors.color4WordType.dispose()
	}
	// Decoration update function
	let updateDecorations = function (bypassLanguageCheck = false, updateColors = false) {
		if( bypassLanguageCheck == false ){
			if((window.activeTextEditor!.document.languageId) !== 'log'){
				return
			}
		}

		if(updateColors === true){
			clearDecorations();
			colors = getColorSettings();
		}

		// Read in document text
		let editorText = window.activeTextEditor!.document.getText();

		// Define object per expression - will be moved to a separate file and be accessible via contributions
		var decorationOptions = [
			{
				name: 'match1',
				expression: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.matchExpression,
				decoration: 'color1'
			},
			{
				name: 'match2',
				expression: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.matchExpression,
				decoration: 'color2'
			},
			{
				name: 'match3',
				expression: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.matchExpression,
				decoration: 'color3'
			},
			{
				name: 'match4',
				expression: workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.matchExpression,
				decoration: 'color4'
			}
		]

		// Loop through each decoration option
		decorationOptions.forEach(option => {
			console.debug('Show-TriggerWords: Processing ' + option.name);

			// Update Highlight Line Range
			let optionLineRanges = [];
			let optionLine;
			let optionLineExpression = new RegExp(('(.+)?(' + option.expression + ')(.+)?'),'gim');
			while (optionLine = optionLineExpression.exec(editorText)) {
				let lineStart = window.activeTextEditor!.document.positionAt(optionLine.index);
				let lineStop = window.activeTextEditor!.document.positionAt(optionLine.index + optionLine[0].length);
				let lineRange = (new Range(lineStart, lineStop));
				optionLineRanges.push(lineRange);
				continue;
			};

			// Update Highlight Word Range
			let optionWordRanges = [];
			let optionWord;
			let optionWordExpression = new RegExp(option.expression,'gim');
			while (optionWord = optionWordExpression.exec(editorText)) {
				let wordStart = window.activeTextEditor!.document.positionAt(optionWord.index);
				let wordStop = window.activeTextEditor!.document.positionAt(optionWord.index + optionWord[0].length);
				let wordRange = (new Range(wordStart, wordStop));
				optionWordRanges.push(wordRange);
				continue;
			};

			// Apply the correct decorations based on the option's decoration value
			switch(option.decoration){
				case 'color1'	:	{
					if(workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.enabled === true){
						window.activeTextEditor!.setDecorations(colors.color1LineType, optionLineRanges);
						window.activeTextEditor!.setDecorations(colors.color1WordType, optionWordRanges);
					};
				};
				case 'color2'	:	{
					if(workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.enabled === true){
						window.activeTextEditor!.setDecorations(colors.color2LineType, optionLineRanges);
						window.activeTextEditor!.setDecorations(colors.color2WordType, optionWordRanges);
					};
				};
				case 'color3'	:	{
					if(workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.enabled === true){
						window.activeTextEditor!.setDecorations(colors.color3LineType, optionLineRanges);
						window.activeTextEditor!.setDecorations(colors.color3WordType, optionWordRanges);
					};
				};
				case 'color4'	:	{
					if(workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.enabled === true){
						window.activeTextEditor!.setDecorations(colors.color4LineType, optionLineRanges);
						window.activeTextEditor!.setDecorations(colors.color4WordType, optionWordRanges);
					};
				};
			}
		});
	};
	// Prevents the updateDecorations function from triggering too often
	var timeout: NodeJS.Timeout;
    function triggerUpdateDecorations() {
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(() => {
            updateDecorations();
        }, 500);
	}
    if (window.activeTextEditor) {
        triggerUpdateDecorations();
    }
	// The user switched to another document
    window.onDidChangeVisibleTextEditors(function (editor) {
        triggerUpdateDecorations();
    }, null, context.subscriptions);
	// Content of document changed
    workspace.onDidChangeTextDocument(function (event) {
        let activeEditor = window.activeTextEditor;
        if (activeEditor && event.document === activeEditor.document) {
            triggerUpdateDecorations();
        }
	}, null, context.subscriptions);
	workspace.onDidChangeConfiguration(function(){
		console.debug('Show-TriggerWords: Configuration Update Detected')
		updateDecorations(false,true)
	});
  	// The user calls the 'Show Trigger Words' command from the command palette
	let disposable = commands.registerCommand('extension.showTriggerWords', () => {
		updateDecorations(true);
	});
	context.subscriptions.push(disposable);
}
export function deactivate() { }

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_1 = require("vscode");
function getColorSettings() {
    // Define the color options - will be moved to a separate file and be accessible via contributions 
    const color1Word = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.matchColor, color: 'black', fontWeight: 'bolder' };
    const color1Line = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.matchColor + '30', isWholeLine: true };
    const color2Word = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.matchColor, color: 'black', fontWeight: 'bolder' };
    const color2Line = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.matchColor + '30', isWholeLine: true };
    const color3Word = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.matchColor, color: 'black', fontWeight: 'bolder' };
    const color3Line = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.matchColor + '30', isWholeLine: true };
    const color4Word = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.matchColor, color: 'black', fontWeight: 'bolder' };
    const color4Line = { backgroundColor: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.matchColor + '30', isWholeLine: true };
    // Build the decorationTypes
    const color1LineType = vscode_1.window.createTextEditorDecorationType(color1Line);
    const color1WordType = vscode_1.window.createTextEditorDecorationType(color1Word);
    const color2LineType = vscode_1.window.createTextEditorDecorationType(color2Line);
    const color2WordType = vscode_1.window.createTextEditorDecorationType(color2Word);
    const color3LineType = vscode_1.window.createTextEditorDecorationType(color3Line);
    const color3WordType = vscode_1.window.createTextEditorDecorationType(color3Word);
    const color4LineType = vscode_1.window.createTextEditorDecorationType(color4Line);
    const color4WordType = vscode_1.window.createTextEditorDecorationType(color4Word);
    return {
        color1LineType,
        color2LineType,
        color3LineType,
        color4LineType,
        color1WordType,
        color2WordType,
        color3WordType,
        color4WordType
    };
}
exports.getColorSettings = getColorSettings;
;
function activate(context) {
    let colors = getColorSettings();
    console.log('mpearon.vsce.show-triggerwords has been initialized');
    // No editor is open, cancel load
    if (!vscode_1.window.activeTextEditor)
        return;
    function clearDecorations() {
        colors.color1LineType.dispose();
        colors.color1WordType.dispose();
        colors.color2LineType.dispose();
        colors.color2WordType.dispose();
        colors.color3LineType.dispose();
        colors.color3WordType.dispose();
        colors.color4LineType.dispose();
        colors.color4WordType.dispose();
    }
    // Decoration update function
    let updateDecorations = function (bypassLanguageCheck = false, updateColors = false) {
        if (bypassLanguageCheck == false) {
            if ((vscode_1.window.activeTextEditor.document.languageId) !== 'log') {
                return;
            }
        }
        if (updateColors === true) {
            clearDecorations();
            colors = getColorSettings();
        }
        // Read in document text
        let editorText = vscode_1.window.activeTextEditor.document.getText();
        // Define object per expression - will be moved to a separate file and be accessible via contributions
        var decorationOptions = [
            {
                name: 'match1',
                expression: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.matchExpression,
                decoration: 'color1'
            },
            {
                name: 'match2',
                expression: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.matchExpression,
                decoration: 'color2'
            },
            {
                name: 'match3',
                expression: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.matchExpression,
                decoration: 'color3'
            },
            {
                name: 'match4',
                expression: vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.matchExpression,
                decoration: 'color4'
            }
        ];
        // Loop through each decoration option
        decorationOptions.forEach(option => {
            console.debug('Show-TriggerWords: Processing ' + option.name);
            // Update Highlight Line Range
            let optionLineRanges = [];
            let optionLine;
            let optionLineExpression = new RegExp(('(.+)?(' + option.expression + ')(.+)?'), 'gim');
            while (optionLine = optionLineExpression.exec(editorText)) {
                let lineStart = vscode_1.window.activeTextEditor.document.positionAt(optionLine.index);
                let lineStop = vscode_1.window.activeTextEditor.document.positionAt(optionLine.index + optionLine[0].length);
                let lineRange = (new vscode_1.Range(lineStart, lineStop));
                optionLineRanges.push(lineRange);
                continue;
            }
            ;
            // Update Highlight Word Range
            let optionWordRanges = [];
            let optionWord;
            let optionWordExpression = new RegExp(option.expression, 'gim');
            while (optionWord = optionWordExpression.exec(editorText)) {
                let wordStart = vscode_1.window.activeTextEditor.document.positionAt(optionWord.index);
                let wordStop = vscode_1.window.activeTextEditor.document.positionAt(optionWord.index + optionWord[0].length);
                let wordRange = (new vscode_1.Range(wordStart, wordStop));
                optionWordRanges.push(wordRange);
                continue;
            }
            ;
            // Apply the correct decorations based on the option's decoration value
            switch (option.decoration) {
                case 'color1':
                    {
                        if (vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match1.enabled === true) {
                            vscode_1.window.activeTextEditor.setDecorations(colors.color1LineType, optionLineRanges);
                            vscode_1.window.activeTextEditor.setDecorations(colors.color1WordType, optionWordRanges);
                        }
                        ;
                    }
                    ;
                case 'color2':
                    {
                        if (vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match2.enabled === true) {
                            vscode_1.window.activeTextEditor.setDecorations(colors.color2LineType, optionLineRanges);
                            vscode_1.window.activeTextEditor.setDecorations(colors.color2WordType, optionWordRanges);
                        }
                        ;
                    }
                    ;
                case 'color3':
                    {
                        if (vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match3.enabled === true) {
                            vscode_1.window.activeTextEditor.setDecorations(colors.color3LineType, optionLineRanges);
                            vscode_1.window.activeTextEditor.setDecorations(colors.color3WordType, optionWordRanges);
                        }
                        ;
                    }
                    ;
                case 'color4':
                    {
                        if (vscode_1.workspace.getConfiguration('vsce-show-triggerwords').highlighting.match4.enabled === true) {
                            vscode_1.window.activeTextEditor.setDecorations(colors.color4LineType, optionLineRanges);
                            vscode_1.window.activeTextEditor.setDecorations(colors.color4WordType, optionWordRanges);
                        }
                        ;
                    }
                    ;
            }
        });
    };
    // Prevents the updateDecorations function from triggering too often
    var timeout;
    function triggerUpdateDecorations() {
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(() => {
            updateDecorations();
        }, 500);
    }
    if (vscode_1.window.activeTextEditor) {
        triggerUpdateDecorations();
    }
    // The user switched to another document
    vscode_1.window.onDidChangeVisibleTextEditors(function (editor) {
        triggerUpdateDecorations();
    }, null, context.subscriptions);
    // Content of document changed
    vscode_1.workspace.onDidChangeTextDocument(function (event) {
        let activeEditor = vscode_1.window.activeTextEditor;
        if (activeEditor && event.document === activeEditor.document) {
            triggerUpdateDecorations();
        }
    }, null, context.subscriptions);
    vscode_1.workspace.onDidChangeConfiguration(function () {
        console.debug('Show-TriggerWords: Configuration Update Detected');
        updateDecorations(false, true);
    });
    // The user calls the 'Show Trigger Words' command from the command palette
    let disposable = vscode_1.commands.registerCommand('extension.showTriggerWords', () => {
        updateDecorations(true);
    });
    context.subscriptions.push(disposable);
}
exports.activate = activate;
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map